File: pyfenstein3d/engine/animation.py
from numpy import interp

class Animation:
    def __init__(self, time: float, on_animate=None, on_animate_end=None):
        self.__total_time = time
        self.__on_animate = on_animate
        self.__on_animate_end = on_animate_end
        self.__time = 0
        self.__factor = 0
        self.__is_animating = False

    @property
    def total_time(self):
        return self.__total_time;

    @property
    def time(self):
        return self.__time;

    @property
    def is_animating(self):
        return self.__is_animating;

    @property
    def factor(self):
        return self.__factor;

    def start(self, at: int = 0):
        self.__time = at
        self.__is_animating = True

    def update(self, delta_time: float):
        if self.__is_animating:
            self.__time += delta_time
            self.__factor =  self.__time / self.__total_time
            self.__factor = self.__factor if self.__factor <= 1 else 1;
            if self.__on_animate is not None:
                self.__on_animate(self.__factor)
            if self.__factor >= 1:
                self.__is_animating = False
                if self.__on_animate_end is not None:
                    self.__on_animate_end()
        else:
            self.__factor = 0

File: engine/animation.go
package engine

// Animation drives a time-based animation with optional callbacks.
type Animation struct {
	totalTime    float64
	time         float64
	factor       float64
	isAnimating  bool
	onAnimate    func(float64)
	onAnimateEnd func()
}

// NewAnimation creates a new Animation with the given duration and optional callbacks.
func NewAnimation(totalTime float64, onAnimate func(float64), onAnimateEnd func()) *Animation {
	return &Animation{
		totalTime:    totalTime,
		onAnimate:    onAnimate,
		onAnimateEnd: onAnimateEnd,
	}
}

func (a *Animation) TotalTime() float64  { return a.totalTime }
func (a *Animation) Time() float64       { return a.time }
func (a *Animation) Factor() float64     { return a.factor }
func (a *Animation) IsAnimating() bool   { return a.isAnimating }

// Start begins the animation, optionally at a given time offset.
func (a *Animation) Start(at float64) {
	a.time = at
	a.isAnimating = true
}

// Update advances the animation by deltaTime seconds.
func (a *Animation) Update(deltaTime float64) {
	if a.isAnimating {
		a.time += deltaTime
		a.factor = a.time / a.totalTime
		if a.factor > 1 {
			a.factor = 1
		}
		if a.onAnimate != nil {
			a.onAnimate(a.factor)
		}
		if a.factor >= 1 {
			a.isAnimating = false
			if a.onAnimateEnd != nil {
				a.onAnimateEnd()
			}
		}
	} else {
		a.factor = 0
	}
}
